### PR TITLE
#3 Implement IPython REPL

### PR DESCRIPTION
**Pull request by anxolerd**
Added blueberrypy ishell command

Implement #3

### CODE REVIEW

**Comment by anxolerd** [comment]
It seems IPython 6.0.0 is not available for Python2.7 Do you plan to support Python2 ?
If no, should I downgrade IPython to lower version?

File: src/blueberrypy/command.py [hunk 1]
```diff
@@ -445,6 +483,8 @@ def get_command_parser_and_callback(command):
             doc, callback = create.__doc__, create
         elif command == "console":
             doc, callback = console.__doc__, console
+        elif command == "ishell":
```
Comment (webknjaz) [inline src/blueberrypy/command.py]: I'd go for extending existing command with autonegotiation of REPL, available for use. See django-extensions - > shell_plus

File: setup.py [hunk 2]
```diff
@@ -10,7 +10,20 @@
                     "docopt>=0.6",
                     "Jinja2>=2.7",
                     "PyYAML>=3.10",
-                    "python-dateutil>=2.2"]
+                    "python-dateutil>=2.2",
+                    # IPython dependencies:
```
Comment (webknjaz) [inline setup.py]: Why would you need to lock recursive deps?

**Review by webknjaz — CHANGES_REQUESTED** [review]

**Comment by webknjaz** [comment]
> should I downgrade

There's environment markers for this purpose

File: setup.py (hunk 2 above)
Comment (anxolerd) [reply setup.py]: Because I do not need any surprises of unexpected updates.
Do you think I should not do that and lock only IPython dependency?

Comment (webknjaz) [reply setup.py]: In a project of a library type you shouldn't restrict versioning strictly unless for corner cases.

File: setup.py [hunk 3]
```diff
@@ -12,18 +12,8 @@
...
-                    "pexpect==4.2.1",
-                    "pickleshare==0.7.4",
-                    "prompt-toolkit==1.0.14",
-                    "ptyprocess==0.5.1",
-                    "Pygments==2.2.0",
-                    "simplegeneric==0.8.1",
-                    "traitlets==4.3.2",
-                    "wcwidth==0.1.7"]
+                    "ipython==5.3.0" if sys.version_info < (3, 3)
+                    else "ipython==6.0.0"]
```
Comment (webknjaz) [inline setup.py]: `>=`

File: setup.py [hunk 4]
```diff
@@ -12,18 +12,8 @@
...
-                    "jedi==0.10.2",
-                    "pexpect==4.2.1",
-                    "pickleshare==0.7.4",
-                    "prompt-toolkit==1.0.14",
-                    "ptyprocess==0.5.1",
-                    "Pygments==2.2.0",
-                    "simplegeneric==0.8.1",
-                    "traitlets==4.3.2",
-                    "wcwidth==0.1.7"]
+                    "ipython==5.3.0" if sys.version_info < (3, 3)
```
Comment (webknjaz) [inline setup.py]: `<=`

File: src/blueberrypy/command.py [hunk 5]
```diff
@@ -1,21 +1,26 @@
...
 import sys
 import re
 import textwrap
 
 from datetime import datetime
 from functools import partial
+from code import InteractiveConsole
 
 import cherrypy
+from IPython.terminal.interactiveshell import TerminalInteractiveShell
```
Comment (webknjaz) [inline src/blueberrypy/command.py]: this will fail if ipython isn't installed

**Review by webknjaz — CHANGES_REQUESTED** [review]
plz also move ipython to extras

**Comment by webknjaz** [comment]
can you borrow the whole logic from django-extensions?

File: setup.py (hunk 2 above)
Comment (anxolerd) [reply setup.py]: Well, I do not trust them enough to do that.

But if you wish I may fix that if you tell me which versions should I allow.

File: src/blueberrypy/command.py (hunk 5 above)
Comment (anxolerd) [reply src/blueberrypy/command.py]: Ipython is in core dependencies so I guess it will be installed with the blueberrypy

**Comment by anxolerd** [comment]
> can you borrow the whole logic from django-extensions?

What do you mean by that?

File: src/blueberrypy/command.py (hunk 5 above)
Comment (webknjaz) [reply src/blueberrypy/command.py]: and it shouldn't be there

**Comment by webknjaz** [comment]
copy-paste driven development, I think

File: setup.py (hunk 2 above)
Comment (webknjaz) [reply setup.py]: see below

File: src/blueberrypy/command.py (hunk 5 above)
Comment (anxolerd) [reply src/blueberrypy/command.py]: Why shouldn't it?

**Comment by anxolerd** [comment]
> copy-paste driven development, I think

I do not understand what exactly you want me to copy-paste. Other shells support? Or what?

File: src/blueberrypy/command.py (hunk 5 above)
Comment (webknjaz) [reply src/blueberrypy/command.py]: because it's not a runtime dependency, it should be optional; if one wants to use ipython they'll install it.

**Comment by anxolerd** [comment]
Updated PR. @webknjaz pls review again =)

File: setup.py [hunk 6]
```diff
@@ -75,7 +75,10 @@
                               "Routes>=2.0",
                               "backlash>=0.0.5",
                               "Shapely>=1.2",
-                              "GeoAlchemy2>=0.2.4"],
+                              "GeoAlchemy2>=0.2.4",
+                              "ipython<=5.3.0" if sys.version_info < (3, 3) else "ipython>=6.0.0"],
```
Comment (webknjaz) [inline setup.py]: This conditional must be computed once and passed via variable; alternatively, you may wrap it into lambda. Copy-pasted code snippets lead to unexpected diversity increase of code base over time.

File: src/blueberrypy/command.py [hunk 7]
```diff
@@ -1,21 +1,25 @@
...
 from functools import partial
+from code import InteractiveConsole
 
 import cherrypy
 from docopt import docopt
 from cherrypy.process import servers
 from cherrypy.process.plugins import Daemonizer, DropPrivileges, PIDFile
 
 import blueberrypy
+import blueberrypy.shell as shell
```
Comment (webknjaz) [inline src/blueberrypy/command.py]: Isn't `from blueberrypy import shell` shorter?

File: src/blueberrypy/shell.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        model.metadata.bind = engine
+        ns['session'] = session = Session()
+        session.bind.echo = True
+        atexit.register(session.close)
+
+    return ns
+
+
+def get_package_name(config):
+    return (
```
Comment (webknjaz) [inline src/blueberrypy/shell.py]: `getattr(config, 'project_metadata', {}).get('package', None)`

File: src/blueberrypy/shell.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        config.project_metadata["package"]
+        if config.project_metadata else None
+    )
+
+
+def _make_sa_engine(config):
```
Comment (webknjaz) [inline src/blueberrypy/shell.py]: Can't this code be shared with SQLAlchemy Tool module?

**Review by webknjaz — CHANGES_REQUESTED** [review]
So what about other shells?

File: src/blueberrypy/shell.py (hunk 9 above)
Comment (anxolerd) [reply src/blueberrypy/shell.py]: No it cannot be shared properly =(

File: src/blueberrypy/shell.py (hunk 8 above)
Comment (anxolerd) [reply src/blueberrypy/shell.py]: I think my version is more explicit.

File: src/blueberrypy/command.py (hunk 7 above)
Comment (anxolerd) [reply src/blueberrypy/command.py]: Yes, it is.

File: setup.py (hunk 6 above)
Comment (anxolerd) [reply setup.py]: Agree

**Comment by anxolerd** [comment]
I think we'll stop on IPython for now. Will add other shells on demand.

**Comment by anxolerd** [comment]
@webknjaz:
- removed code duplication from setup.py
- shorter import is used

I do not see any good way to share code with SQLAlchemy tool without overcomplicating the tool itself.

File: src/blueberrypy/shell.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+                ns[name] = getattr(app_module, name)
+        else:
+            for name, obj in vars(app_module).viewitems():
+                if not name.startswith("_"):
+                    ns[name] = obj
+
+    ns['cherrypy'] = importlib.import_module("cherrypy")
+    ns['blueberrypy'] = importlib.import_module("blueberrypy")
+
+    if config.use_sqlalchemy:
```
Comment (webknjaz) [inline src/blueberrypy/shell.py]: I think you could reuse this property https://github.com/open-craft-guild/blueberrypy/blob/master/src/blueberrypy/config.py#L270

File: src/blueberrypy/shell.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
+import atexit
+import importlib
+
+from blueberrypy.config import BlueberryPyConfiguration
+
+
+def get_user_namespace(config, include_pkg=False):
+    assert isinstance(config, BlueberryPyConfiguration), type(config)
```
Comment (webknjaz) [inline src/blueberrypy/shell.py]: `type(config)` doesn't look like sufficient error message

File: src/blueberrypy/shell.py (hunk 10 above)
Comment (anxolerd) [reply src/blueberrypy/shell.py]: The logic I used was present in old blueberrypy console. This is why I kept it.

File: src/blueberrypy/shell.py (hunk 11 above)
Comment (anxolerd) [reply src/blueberrypy/shell.py]: Agree. Fixed in 44e80e3

**Review by webknjaz — APPROVED** [review]

**Comment by anxolerd** [comment]
May I merge?

**Comment by anxolerd** [comment]
@webknjaz can this be merged, or any further changes are required?

**Comment by webknjaz** [comment]
Go ahead and merge it.

**Comment by anxolerd** [comment]
@webknjaz I am not authorized to merge this pull request.

**Comment by webknjaz** [comment]
> This branch is out of sync with the base branch.
Changes from the base branch must be synced with this branch before merging.

Fix this first

**Pull request closed without merging**